src/main.cc

```diff
@@ -1,14 +1,175 @@
+
 #include "nan.h"
+#include <windows.h>
+#include <stdio.h>
+#include <tchar.h>
+
+using namespace Nan;
+using namespace v8;
+
+#define UNICODE
+
+#define MAX_KEY_LENGTH 255
+#define MAX_VALUE_NAME 16383
 
 namespace {
 
-NAN_METHOD(HelloWorld) {
-  auto value = "Hello world!";
-  info.GetReturnValue().Set(Nan::New(value).ToLocalChecked());
+v8::Local<v8::Object> CreateEntry(TCHAR *name, TCHAR *type, char *data) {
+  auto obj = Nan::New<v8::Object>();
+  obj->Set(Nan::New("name").ToLocalChecked(), Nan::New(name).ToLocalChecked());
+  obj->Set(Nan::New("type").ToLocalChecked(), Nan::New(type).ToLocalChecked());
+  obj->Set(Nan::New("data").ToLocalChecked(), Nan::New<v8::String>(data).ToLocalChecked());
+  return obj;
+}
+
+v8::Local<v8::Object> CreateEntry(TCHAR *name, TCHAR *type, DWORD data) {
+  auto obj = Nan::New<v8::Object>();
+  obj->Set(Nan::New("name").ToLocalChecked(), Nan::New(name).ToLocalChecked());
+  obj->Set(Nan::New("type").ToLocalChecked(), Nan::New(type).ToLocalChecked());
+  obj->Set(Nan::New("data").ToLocalChecked(), Nan::New((INT32)data));
+  return obj;
+}
+
+NAN_METHOD(ReadValues) {
+  if (info.Length() < 2) {
+    Nan::ThrowTypeError("Wrong number of arguments");
+    return;
+  }
+
+  if (!info[0]->IsNumber()) {
+    Nan::ThrowTypeError("A number was expected for the first argument, but wasn't received.");
+    return;
+  }
+
+  if (!info[1]->IsString()) {
+    Nan::ThrowTypeError("A string was expected for the second argument, but wasn't received.");
+    return;
+  }
+
+  ULONG first = info[0]->NumberValue();
+  auto second = *v8::String::Utf8Value(info[1]);
+
+  HKEY hCurrentKey;
+  LONG openKey;
+
+  openKey = RegOpenKeyEx(
+    (HKEY)first,
+    second,
+    0,
+    KEY_READ | KEY_WOW64_64KEY,
+    &hCurrentKey);
+
+  if (openKey != ERROR_SUCCESS) {
+    char* errorMessage = NULL;
+    sprintf(errorMessage, "RegOpenKeyEx failed - exit code: '%d'", openKey);
+    Nan::ThrowTypeError(errorMessage);
+    return;
+  }
+
+  TCHAR achClass[MAX_PATH] = TEXT("");  // buffer for class name
+  DWORD cchClassName = MAX_PATH;        // size of class string
+  DWORD cSubKeys = 0;                   // number of subkeys
+  DWORD cbMaxSubKey;                    // longest subkey size
+  DWORD cchMaxClass;                    // longest class string
+  DWORD cValues;                        // number of values for key
+  DWORD cchMaxValue;                    // longest value name
+  DWORD cbMaxValueData;                 // longest value data
+
+  TCHAR achValue[MAX_VALUE_NAME];
+  DWORD cchValue = MAX_VALUE_NAME;
+
+  auto retCode = RegQueryInfoKey(
+    hCurrentKey,
+    achClass,
+    &cchClassName,
+    NULL, // reserved
+    &cSubKeys,
+    &cbMaxSubKey,
+    &cchMaxClass,
+    &cValues,
+    &cchMaxValue,
+    &cbMaxValueData,
+    NULL, // can ignore these values
+    NULL);
+
+  if (retCode != ERROR_SUCCESS) {
+    char* errorMessage = NULL;
+    sprintf(errorMessage, "RegQueryInfoKey failed - exit code: '%d'", retCode);
```

Comment:
That's an error. `errorMessage` is supposed to be a buffer of appropriate size, not a `nullptr`.

Reply:
@danilaml are you interested in submitting a fix? I'm currently trying to wrap my head around the encodings being served out of the registry APIs and marshalling them correctly to the caller.

Reply:
@shiftkey ok. Btw, WinAPI is usually UTF-16 (UCS-2 to be precise), unless specified otherwise (there are legacy win95/98 compat functions that deal in ascii ending in A and some that understand UTF8).

Reply:
Yeah, I opened #17 to track this

Reply:
@shiftkey btw, are there any min compiler requirements? I.e. is using some basic C++11 features fine (there are some memory leaks here that can be easily fixed by using smart pointers, though it's not hard fixing them manually)?

Reply:
@danilaml the node toolchain targets MSVS 2015, so modern C++ is totally fine

Reply:
@danilaml there's also CI already setup, so I'm pretty sure we've got a good harness for testing whatever you might want to try out